src/OpenTelemetry.Instrumentation.AspNet/Implementation/HttpInListener.cs

```diff
@@ -206,9 +207,36 @@ public override void OnStopActivity(Activity activity, object payload)
 
                 if (!string.IsNullOrEmpty(template))
                 {
-                    // Override the name that was previously set to the path part of URL.
-                    activityToEnrich.DisplayName = template;
-                    activityToEnrich.SetTag(SemanticConventions.AttributeHttpRoute, template);
+                    bool isRootPath = request.Path == "/";
+                    bool isIdOptional = false;
+
+                    // For urls without ID, for e.g. GET all resources of a kind (/api/Values/), we don't need to use template information
+                    if (!isRootPath && request.Path.Split(new char[] { '/' }, StringSplitOptions.RemoveEmptyEntries).Length <
```

Comment:
@utpilla Can you explain this logic a bit? I just want to double-check we're not doing something that will only work with the default routes as users can always modify them 😄

Reply:
This is to check if the request doesn't need any Id. For eg. if the request that we are looking at is something like "api/books/" instead of "api/books/1",  we want to set the Activity DisplayName as "api/books" and not "api/books/{id}. I'm doing that by comparing the count of the path segments of the actual request vs template. If the actual request has lesser number of path segments than the template, that implies that it doesn't have an Id so we can just use the displayName that was set by the OnStartActivity method.

It should work with custom route templates as well because we only need to find if a request returns data for a specific Id or does it not need the Id at all which can be found out by comparing the count of path segments.

Reply:
I noticed that if the default route template is configured as:
```
config.Routes.MapHttpRoute(
                name: "DefaultApi",
                routeTemplate: "api/{controller}/{id}",
                defaults: new { controller = "Books", id = RouteParameter.Optional }
```
Then if the request is "host/api/" it will route to "api/books". With the logic implemented in this PR the Activity DisplayName would be "api/" and not "api/books".

Reply:
What if I had a route like /books/1234/author? I like us switching {controller} & {action} to the values because those should generally be well-defined, but doing anything more I think we open ourselves up to trouble 😄

An idea I just had... What if we added a func to options we use to convert template + route data to display name? The default version would be the logic we have here. But if someone wanted to customize it, they could replace the func on options to do something more custom.

Reply:
If the route was /books/1234/author, the Activity DsiplayName would be set to "books/{Id}/author". The template for this could have either been "{controller}/{id}/{action}" or "{controller}/{id}/author". It should set the display name to "books/{Id}/author" for both the cases.

The options idea sounds great!